cluster-autoscaler/simulator/basic_cluster_snapshot_test.go

```diff
@@ -0,0 +1,334 @@
+					for i := 0; i < b.N; i++ {
+						err = clusterSnapshot.Fork()
+						if err != nil {
+							assert.NoError(b, err)
+						}
+						err = clusterSnapshot.AddNode(tmpNode)
+						if err != nil {
+							assert.NoError(b, err)
+						}
+						err = clusterSnapshot.Revert()
+						if err != nil {
+							assert.NoError(b, err)
+						}
+					}
+				})
+			}
+		}
+	}
+}
+
+type snapshotFactory func() ClusterSnapshot
+
+func TestFork(t *testing.T) {
+	nodeCount := 3
+	podCount := 90
+
+	nodes := createTestNodes(nodeCount)
+	extraNodes := createTestNodesWithPrefix("tmp", 2)
+	pods := createTestPods(podCount)
+	assignPodsToNodes(pods, nodes)
+
+	snapshots := map[string]snapshotFactory{
+		"basic": func() ClusterSnapshot { return NewBasicClusterSnapshot() },
+		"delta": func() ClusterSnapshot { return NewDeltaClusterSnapshot() },
+	}
+
+	for name, snapshotFactory := range snapshots {
+		t.Run(fmt.Sprintf("%s: fork should not affect base data: adding nodes", name),
+			func(t *testing.T) {
+				clusterSnapshot := snapshotFactory()
+				_ = clusterSnapshot.AddNodes(nodes)
+				_ = clusterSnapshot.Fork()
+				for _, node := range extraNodes {
+					_ = clusterSnapshot.AddNode(node)
+				}
+				forkNodes := clusterSnapshot.GetAllNodes()
+				assert.Equal(t, nodeCount+len(extraNodes), len(forkNodes))
+				_ = clusterSnapshot.Revert()
+				baseNodes := clusterSnapshot.GetAllNodes()
+				assert.Equal(t, nodeCount, len(baseNodes))
+			})
+		t.Run(fmt.Sprintf("%s: fork should not affect base data: adding pods", name),
+			func(t *testing.T) {
+				clusterSnapshot := snapshotFactory()
+				_ = clusterSnapshot.AddNodes(nodes)
+				_ = clusterSnapshot.Fork()
+				for _, pod := range pods {
+					_ = clusterSnapshot.AddPod(pod, pod.Spec.NodeName)
+				}
+				forkPods := clusterSnapshot.GetAllPods()
+				assert.Equal(t, len(pods), len(forkPods))
+				_ = clusterSnapshot.Revert()
+				basePods := clusterSnapshot.GetAllPods()
+				assert.Equal(t, 0, len(basePods))
+			})
+	}
+
+}
+
+func BenchmarkBuildNodeInfoList(b *testing.B) {
+	testCases := []struct {
+		nodeCount int
+	}{
+		{
+			nodeCount: 1000,
+		},
+		{
+			nodeCount: 5000,
+		},
+		{
+			nodeCount: 15000,
+		},
+		{
+			nodeCount: 100000,
+		},
+	}
+
+	for _, tc := range testCases {
+		b.Run(fmt.Sprintf("fork add 1000 to %d", tc.nodeCount), func(b *testing.B) {
+			nodes := createTestNodes(tc.nodeCount + 1000)
+			snapshot := NewDeltaClusterSnapshot()
+			if err := snapshot.AddNodes(nodes[:tc.nodeCount]); err != nil {
+				assert.NoError(b, err)
+			}
+			if err := snapshot.Fork(); err != nil {
+				assert.NoError(b, err)
+			}
+			if err := snapshot.AddNodes(nodes[tc.nodeCount:]); err != nil {
+				assert.NoError(b, err)
+			}
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				list := snapshot.data.buildNodeInfoList()
+				if len(list) != tc.nodeCount+1000 {
+					assert.Equal(b, len(list), tc.nodeCount+1000)
+				}
+			}
+		})
+	}
+	for _, tc := range testCases {
```

Comment:
Seems like the shared part is minimal - maybe split into two functions?

Reply:
Done